Avoid requiring the latest Go toolchain patch version to build

joestringer opened the pull request:
Forcing developers to change their local toolchain within patch releases
introduces more problems than it's worth. Remove the automatic go.mod
update and don't require newer patch releases of the latest Go version.

Related: https://github.com/cilium/cilium/pull/27820
Related: https://github.com/golang/go/issues/63522

aanm commented on go.mod:
```diff
@@ -1,7 +1,6 @@
 module github.com/cilium/cilium
 
-// renovate: datasource=golang-version depName=go
```
I think we can keep this renovate line and add the following (not tested):
```
$ git diff
diff --git a/.github/renovate.json5 b/.github/renovate.json5
index a1544a26f8..8170b98d68 100644
--- a/.github/renovate.json5
+++ b/.github/renovate.json5
@@ -119,6 +119,26 @@
         "main"
       ]
     },
+    {
+      "groupName": "only major minor golang",
+      "groupSlug": "only-major-minor-golang",
+      "matchFiles": [
+        "go.mod",
+      ],
+      "matchDepNames": [
+        "go"
+      ],
+      "matchDatasources": [
+        "golang-version"
+      ],
+      "matchUpdateTypes": [
+        "major",
+        "minor",
+      ],
+      matchBaseBranches: [
+        "main"
+      ]
+    },
     {
       // Do not allow any updates into stable branches.
       "enabled": false,
```

joestringer replied:
SGTM. Does this generate `1.21.0` as output?

joestringer replied:
Oh incidentally, I just took this patch verbatim and applied it... How would one test this?

aanm replied:
> SGTM. Does this generate 1.21.0 as output?

You can test it locally with https://docs.renovatebot.com/modules/platform/local/

joestringer replied:
LGTM, as long as the lack of GitHub token doesn't break local testing:

```
$ docker run -ti -v $(pwd):/usr/src/app docker.io/renovate/renovate:full renovate --platform=local
 INFO: Repository started (repository=local)
       "renovateVersion": "37.31.1"
fatal: detected dubious ownership in repository at '/usr/src/app'
To add an exception for this directory, call:

        git config --global --add safe.directory /usr/src/app
fatal: detected dubious ownership in repository at '/usr/src/app'
To add an exception for this directory, call:

        git config --global --add safe.directory /usr/src/app
fatal: detected dubious ownership in repository at '/usr/src/app'
To add an exception for this directory, call:

        git config --global --add safe.directory /usr/src/app
fatal: detected dubious ownership in repository at '/usr/src/app'
To add an exception for this directory, call:

        git config --global --add safe.directory /usr/src/app
fatal: detected dubious ownership in repository at '/usr/src/app'
To add an exception for this directory, call:

        git config --global --add safe.directory /usr/src/app
fatal: detected dubious ownership in repository at '/usr/src/app'
To add an exception for this directory, call:

        git config --global --add safe.directory /usr/src/app
 INFO: Dependency extraction complete (repository=local)
       "stats": {
         "managers": {
           "dockerfile": {"fileCount": 9, "depCount": 31},
           "github-actions": {"fileCount": 49, "depCount": 496},
           "gomod": {"fileCount": 1, "depCount": 252},
           "kubernetes": {"fileCount": 1, "depCount": 2},
           "regex": {"fileCount": 42, "depCount": 121}
         },
         "total": {"fileCount": 102, "depCount": 902}
       }
 WARN: GitHub token is required for some dependencies (repository=local)
       "githubDeps": [
         "actions/checkout",
         "docker/setup-buildx-action",
         "docker/setup-qemu-action",
         "sigstore/cosign-installer",
         "docker/login-action",
         "docker/build-push-action",
         "actions/upload-artifact",
         "cilium/actions-app-token",
         "actions/download-artifact",
         "actions/cache",
         "actions/stale",
         "myrotvorets/set-commit-status-action",
         "cilium/cilium-cli",
         "azure/login",
         "aanm/junit2md",
         "aws-actions/configure-aws-credentials",
         "helm/kind-action",
         "cilium/little-vm-helper",
         "google-github-actions/auth",
         "google-github-actions/setup-gcloud",
         "nick-invision/retry",
         "actions/setup-go",
         "cilium/cilium",
         "anchore/scan-action",
         "dorny/paths-filter",
         "actions/github-script",
         "KyleMayes/install-llvm-action",
         "github/codeql-action",
         "golangci/golangci-lint-action",
         "actions/setup-python",
         "hramos/needs-attention",
         "cilium/reusable-workflows",
         "google/oss-fuzz",
         "cilium/hubble",
         "google/gops",
         "kubernetes-sigs/bom",
         "eksctl-io/eksctl",
         "kubernetes/kubernetes",
         "kubernetes-sigs/kind",
         "kubernetes/kops"
       ]
 INFO: Repository finished (repository=local)
       "cloned": undefined,
       "durationMs": 171608
```

joestringer replied:
I have no idea what this output means. I managed to get a local copy built and running and it got rid of all the warnings (`$GITHUB_COM_TOKEN`) but it's not really telling me anything. I can't tell if that means:
- Config is good, success because there's no errors
- Config is bad, command succeeds but the configuration is ineffectual.

```
cilium $ renovate --platform=local
 INFO: Repository started (repository=local)
       "renovateVersion": "0.0.0-semantic-release"
 INFO: Dependency extraction complete (repository=local)
       "stats": {
         "managers": {
           "dockerfile": {"fileCount": 9, "depCount": 31},
           "github-actions": {"fileCount": 49, "depCount": 499},
           "gomod": {"fileCount": 1, "depCount": 254},
           "kubernetes": {"fileCount": 1, "depCount": 2},
           "regex": {"fileCount": 42, "depCount": 122}
         },
         "total": {"fileCount": 102, "depCount": 908}
       }
 INFO: Repository finished (repository=local)
       "cloned": undefined,
       "durationMs": 30337
```

If I use `--dry-run=full` then it fails out with what looks like an upstream renovate bug:

```
cilium $ renovate --platform=local --dry-run=full
 INFO: Repository started (repository=local)
       "renovateVersion": "0.0.0-semantic-release"
 INFO: Dependency extraction complete (repository=local)
       "stats": {
         "managers": {
           "dockerfile": {"fileCount": 9, "depCount": 31},
           "github-actions": {"fileCount": 49, "depCount": 499},
           "gomod": {"fileCount": 1, "depCount": 254},
           "kubernetes": {"fileCount": 1, "depCount": 2},
           "regex": {"fileCount": 42, "depCount": 122}
         },
         "total": {"fileCount": 102, "depCount": 908}
       }
 INFO: DRY-RUN: Would ensure Dependency Dashboard (repository=local)
       "title": "Dependency Dashboard"
 INFO: DRY-RUN: Would save repository cache. (repository=local)
ERROR: Repository has unknown error (repository=local)
       "err": {
         "message": "Cannot convert undefined or null to object",
         "stack": "TypeError: Cannot convert undefined or null to object\n    at Function.keys (<anonymous>)\n    at getBranchList (/home/joe/git/renovate/lib/util/git/index.ts:576:17)\n    at pruneStaleBranches (/home/joe/git/renovate/lib/workers/reposit
ory/finalize/prune.ts:112:39)\n    at finalizeRepo (/home/joe/git/renovate/lib/workers/repository/finalize/index.ts:21:27)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at Object.renovateRepository (/home/joe/git/renovate
/lib/workers/repository/index.ts:94:7)\n    at attributes.repository (/home/joe/git/renovate/lib/workers/global/index.ts:184:11)\n    at start (/home/joe/git/renovate/lib/workers/global/index.ts:169:7)\n    at /home/joe/git/renovate/lib/renovate.ts:18:22"
       }
 INFO: Repository finished (repository=local)
       "cloned": undefined,
       "durationMs": 47498
 INFO: Renovate is exiting with a non-zero code due to the following logged errors
       "loggerErrors": [
         {
           "name": "renovate",
           "level": 50,
           "logContext": "rs4MDLc4tifxh-rQL9lMT",
           "repository": "local",
           "err": {
             "message": "Cannot convert undefined or null to object",
             "stack": "TypeError: Cannot convert undefined or null to object\n    at Function.keys (<anonymous>)\n    at getBranchList (/home/joe/git/renovate/lib/util/git/index.ts:576:17)\n    at pruneStaleBranches (/home/joe/git/renovate/lib/workers/rep
ository/finalize/prune.ts:112:39)\n    at finalizeRepo (/home/joe/git/renovate/lib/workers/repository/finalize/index.ts:21:27)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at Object.renovateRepository (/home/joe/git/reno
vate/lib/workers/repository/index.ts:94:7)\n    at attributes.repository (/home/joe/git/renovate/lib/workers/global/index.ts:184:11)\n    at start (/home/joe/git/renovate/lib/workers/global/index.ts:169:7)\n    at /home/joe/git/renovate/lib/renovate.ts:18
:22"
           },
           "msg": "Repository has unknown error"
         }
       ]
```

joestringer replied:
OK, renovate fails the same way with `--dry-run=full` on tip of main so that gives no signal.

joestringer replied:
I feel like I'm no further to understanding whether the config is correct or not. We still have to just YOLO merge it and hope it's correct?

tklauser replied:
The way I usually test renovate changes is on my private fork with renovate enabled. Then I push the changes to test to the `main` branch of the fork and wait for renovate to open the corresponding PR (or check the renovate dependency dashboard issue for any errors).

joestringer replied:
How did you set up renovate on a private fork? My account doesn't show up under the marketplace search box if I try to add it to my personal account: https://github.com/marketplace/renovate .

joestringer replied:
Aha, that link doesn't allow personal accounts. But this link does: https://github.com/apps/renovate . I've kicked off a test on my fork.

joestringer replied:
OK, tested and the new version works as expected - basically the other existing rule _enables_ all go updates, so we need another rule than then disables patches on top.

joestringer replied:
Tested by going back to 1.20.0 => Renovate proposes 1.21.x . Also reverting to 1.21.0 as per this PR, renovate does not propose an update.

joestringer committed:
go.mod: Set minimal toolchain to v1.21.0

Nothing in the tree depends on newer patch releases of Go, so we can
keep this to the first patch release of this series.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
.github: Only update go.mod for Go minor releases

The goal here is to allow developer toolchains to have more flexibility
around the specific version of Go compared with the latest patch releases.
Note that all of the real-world code and all tests will continue to be
updated to use the latest Go versions, it's just the go.mod that will
start to drift, since the code can actually compile with earlier
versions of Go; we don't require specific patch versions based on the
set of standard library features used in the code.

Co-authored-by: André Martins <andre@cilium.io>
Signed-off-by: Joe Stringer <joe@cilium.io>

tklauser reviewed (approved):
Thanks

chancez commented on go.mod:
```diff
@@ -1,7 +1,7 @@
 
 // renovate: datasource=golang-version depName=go
-go 1.21.3
+go 1.21.0
```
Should this be `1.21` instead of `1.21.0`?

joestringer replied:
No, according to upstream, `go 1.21` means development versions of Go ([context](https://github.com/golang/go/issues/62278#issuecomment-1693538776)) whereas `go 1.21.0` means to require 1.21.0 or newer stable releases ([context](https://github.com/golang/go/issues/63522#issuecomment-1760201533)).
